MINOR. implement --expose-ports option in ducker-ak

soondenana opened the pull request:
This change adds a command line option to the `ducker-ak up` command to
enable exposing ports from docker containers. The exposed ports will be
mapped to the ephemeral ports on the host. The option is called
`expose-ports` and can take either a single value (like 5005) or a range
(like 5005-5009). This port will then exposed from each docker container
that ducker-ak sets up.

Here is one command line example:

> ./docker/ducker-ak up --expose-ports 5005

After doing this, docker will report mapped ports. Here is example
output of `docker ps` command:

CONTAINER ID        IMAGE                             COMMAND                  CREATED             STATUS                PORTS                     NAMES
d470b708391d        ducker-ak-openjdk-8               "/bin/sh -c 'sudo se…"   16 minutes ago      Up 16 minutes         0.0.0.0:32845->5005/tcp   ducker14
4a25ae952f28        ducker-ak-openjdk-8               "/bin/sh -c 'sudo se…"   16 minutes ago      Up 16 minutes         0.0.0.0:32844->5005/tcp   ducker13
d7b8dad9259d        ducker-ak-openjdk-8               "/bin/sh -c 'sudo se…"   16 minutes ago      Up 16 minutes         0.0.0.0:32843->5005/tcp   ducker12
...

Tested by launching a bin/kafka-topics.sh command and then verifying
that I can attach the debugger successfully to the launched binary.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

soondenana committed:
MINOR. Expose a port from each docker container

This change adds a command line option to the `ducker-ak up' command to
enable exposing ports from docker containers. The exposed ports will be
mapped to the ephemeral ports on the host. The option is called
`expose-ports' and can take either a single value (like 5005) or a range
(like 5005-5009). This port will then exposed from each docker container
that ducker-ak sets up.

Here is one command line example:

> ./docker/ducker-ak up --expose-ports 5005

After doing this, docker will report mapped ports. Here is example
output of `docker ps' command:

CONTAINER ID        IMAGE                             COMMAND                  CREATED             STATUS                PORTS                     NAMES
d470b708391d        ducker-ak-openjdk-8               "/bin/sh -c 'sudo se…"   16 minutes ago      Up 16 minutes         0.0.0.0:32845->5005/tcp   ducker14
4a25ae952f28        ducker-ak-openjdk-8               "/bin/sh -c 'sudo se…"   16 minutes ago      Up 16 minutes         0.0.0.0:32844->5005/tcp   ducker13
d7b8dad9259d        ducker-ak-openjdk-8               "/bin/sh -c 'sudo se…"   16 minutes ago      Up 16 minutes         0.0.0.0:32843->5005/tcp   ducker12
...

Tested by launching a bin/kafka-topics.sh command and then verifying
that I can attach the debugger successfully to the launched binary.

soondenana commented:
CC: @cmccabe, @hachikuji, @jsancio, @mumrah

jsancio commented on tests/docker/ducker-ak:
```diff
@@ -70,6 +70,10 @@ up [-n|--num-nodes NUM_NODES] [-f|--force] [docker-image]
     directory should be the ducktape git repo, not the ducktape installed module directory.
 
+    if --expose-ports is specified then we will expose those ports to random ephemeral ports
+    on the host. The argument can be a single port (like 5005) or a port range like (5005-5009).
```
Why a random port? For my own edification, how are you planning to use this?

soondenana replied:
Our systest launch 14 docker containers (used to setup a kafka cluter with ZK etc). Then systest runs using these containers. When things go wrong there is no way to debug those. Exposing ports help us attach debugger and figure out what is going on.

Random ports as we are going to expose same port from all container. If you look at port column above in the commit description, you will see how it pans out.

jsancio replied:
Thanks for the background. How do we expose multiple ports that are not in a range?

cmccabe replied:
It seems like there could be some value in allowing multiple `--expose-ports` options to be specified.  However, we could also leave this as a future improvement.

jsancio replied:
Yep. @soondenana and I talked offline.

soondenana replied:
Done. Now it takes comma separated list of ports to expose.

cmccabe commented:
Thanks for this, @soondenana.  Can you add some text to the help message for --expose-ports giving using a debugger as an example?

jsancio reviewed (approved):
LGTM!

soondenana committed:
Took care of review comments

Following changes were made:

1. Added option to provide list of ports to expose, and
2. Updated documentation to provide an example on how exposed ports can
   be used to perform remote debugging.

soondenana commented:
Updated the documentation as @cmccabe suggested.

And as @jsancio and @cmccabe asked, now user can specify list of comma separated ports to expose. The list of ports can be individual ports or a range.

soondenana committed:
Clarify network mapping in documentation

Added detail on how to look for mapping of exposed ports

cmccabe commented:
Thanks for the revisions, @soondenana.

I'm sorry to be picky about this, but I feel like having so much text here might confuse new users.  It would be better to just create a ./tests/docker/README.md file and put it in Then, we can just specify what the flag does here and say, for an example of setting up a debugger, see the README.md.  That lets us go into a lot of detail in the README.md.

soondenana committed:
Moved detailed documentation to readme.md

Added a note in the description of the --expose-ports option that more
detail can be found in the readme.

soondenana commented:
> Thanks for the revisions, @soondenana.
> 
> I'm sorry to be picky about this, but I feel like having so much text here might confuse new users. It would be better to just create a ./tests/docker/README.md file and put it in Then, we can just specify what the flag does here and say, for an example of setting up a debugger, see the README.md. That lets us go into a lot of detail in the README.md.

I think its reasonable to move the documentation out. I was also thinking that it was too much for the command line. I have however updated the existing README.md file in `tests` directory instead of adding a new one. Otherwise too many places for someone to look for information. Let me know what you think. Here is how it looks like:

![ducker_expose_ports_doc](https://user-images.githubusercontent.com/50422828/64207474-8f198d00-ce51-11e9-8a42-263709291621.png)

cmccabe commented:
LGTM.  Thanks, @soondenana
